File: examples/dekita/.claude/hooks/pr_issue_alignment_check.py
#!/usr/bin/env python3
# - 責務: PR作成時に対象Issueの受け入れ条件を検証
# - 重複なし: closes-validation.pyはキーワード検証、このフックは内容検証
# - 警告型: 未完了の受け入れ条件がある場合、systemMessageで強い警告
"""
PreToolUse hook to verify Issue acceptance criteria when creating PRs.

When `gh pr create` is detected with Closes/Fixes keywords,
this hook:
1. Fetches the Issue content
2. Extracts acceptance criteria (checkbox items in body)
3. Warns if any criteria are incomplete

Background (Issue #543, #592):
- Issue #538 was closed with a PR that implemented something different
- Issue #590 was closed with a PR that only added debug logs (not a fix)
This hook ensures acceptance criteria are visible and verified.
"""

import json
import re
import subprocess

from lib.constants import TIMEOUT_MEDIUM
from lib.execution import log_hook_execution
from lib.results import print_continue_and_log_skip
from lib.session import create_hook_context, parse_hook_input
from lib.strings import strip_quoted_strings

# Maximum length for issue body display (truncated if longer)
MAX_ISSUE_BODY_LENGTH = 1000


def extract_acceptance_criteria(body: str) -> list[tuple[bool, str]]:
    """Extract acceptance criteria (checkbox items) from Issue body.

    Args:
        body: The Issue body text.

    Returns:
        List of (is_completed, criteria_text) tuples.
    """
    criteria = []

    # Match checkbox items: - [ ] or - [x] or - [X]
    # Also handles * [ ] format
    pattern = r"^[\s]*[-*]\s*\[([ xX])\]\s*(.+)$"

    for line in body.split("\n"):
        match = re.match(pattern, line)
        if match:
            is_completed = match.group(1).lower() == "x"
            criteria_text = match.group(2).strip()
            criteria.append((is_completed, criteria_text))

    return criteria


def format_acceptance_criteria_message(
    issue_num: str, title: str, criteria: list[tuple[bool, str]], is_closed: bool = False
) -> str:
    """Format status message for Issue acceptance criteria.

    Includes both incomplete (warning) and completed criteria.

    Args:
        issue_num: The issue number.
        title: The issue title.
        criteria: List of (is_completed, text) tuples.
        is_closed: Whether the issue is already closed.

    Returns:
        Formatted status message summarizing acceptance criteria.
    """
    incomplete = [text for is_completed, text in criteria if not is_completed]
    completed_items = [text for is_completed, text in criteria if is_completed]

    header = f"### Issue #{issue_num}: {title}"
    if is_closed:
        header += " (CLOSED)"
    lines = [header]
    lines.append("")

    if is_closed and incomplete:
        lines.append("ℹ️ *このIssueは既にクローズ済みです。`Closes #N` は効果がありません。*")
        lines.append("")

    if incomplete:
        lines.append(f"❌ **未完了の受け入れ条件: {len(incomplete)}件**")
        for text in incomplete:
            lines.append(f"  - [ ] {text}")
        lines.append("")

    if completed_items:
        lines.append(f"✅ 完了済み: {len(completed_items)}件")
        for text in completed_items:
            lines.append(f"  - [x] {text}")

    return "\n".join(lines)


def extract_issue_numbers_from_body(command: str) -> list[str]:
    """Extract issue numbers from Closes/Fixes keywords in PR body.

    Args:
        command: The bash command string.

    Returns:
        List of issue numbers found.
    """
    body = None

    # Try HEREDOC pattern first (most common in this project)
    # Matches: --body "$(cat <<'EOF' ... EOF )"
    heredoc_match = re.search(
        r'--body\s+"\$\(cat\s+<<[\'"]?EOF[\'"]?\s*(.*?)\s*EOF\s*\)"',
        command,
        re.DOTALL,
    )
    if heredoc_match:
        body = heredoc_match.group(1)

    # Try double-quoted body (may contain escaped quotes)
    if body is None:
        dq_match = re.search(r'--body\s+"((?:[^"\\]|\\.)*)"', command)
        if dq_match:
            body = dq_match.group(1)

    # Try single-quoted body (may contain any chars except single quote)
    if body is None:
        sq_match = re.search(r"--body\s+'([^']*)'", command)
        if sq_match:
            body = sq_match.group(1)

    if body is None:
        return []

    # Find Closes #XXX, Fixes #XXX, Resolves #XXX patterns
    # Case insensitive, handles multiple issues, allows optional colon (e.g., "Closes: #123")
    pattern = r"(?:closes?|fix(?:es)?|resolves?):?\s+#(\d+)"
    matches = re.findall(pattern, body, re.IGNORECASE)

    return list(set(matches))  # Remove duplicates


def is_pr_create_command(command: str) -> bool:
    """Check if command is gh pr create.

    Args:
        command: The bash command string.

    Returns:
        True if this is a gh pr create command.
    """
    cmd = strip_quoted_strings(command)
    return bool(re.search(r"gh\s+pr\s+create\b", cmd))


def fetch_issue_content(issue_number: str) -> tuple[bool, str, str, str]:
    """Fetch issue title, body and state using gh CLI.

    Args:
        issue_number: The issue number.

    Returns:
        Tuple of (success, title, body, state).
        state is "OPEN" or "CLOSED".
    """
    try:
        result = subprocess.run(
            [
                "gh",
                "issue",
                "view",
                issue_number,
                "--json",
                "title,body,state",
            ],
            capture_output=True,
            text=True,
            timeout=TIMEOUT_MEDIUM,
            check=False,
        )

        if result.returncode != 0:
            return (False, "", "", "")

        # Parse JSON output directly instead of using jq
        data = json.loads(result.stdout)
        title = data.get("title") or ""
        body = data.get("body") or ""  # Handle null body (issues with no description)
        state = data.get("state") or "OPEN"
        return (True, title, body, state)

    except (subprocess.TimeoutExpired, OSError, json.JSONDecodeError):
        return (False, "", "", "")


def main():
    """
    PreToolUse hook for Bash commands.

    Displays Issue content when creating PRs with Closes/Fixes keywords.
    """
    result = {"decision": "approve"}

    try:
        input_data = parse_hook_input()
        # Issue #2607: Create context for session_id logging
        ctx = create_hook_context(input_data)
        tool_name = input_data.get("tool_name", "")

        if tool_name != "Bash":
            print_continue_and_log_skip(
                "pr-issue-alignment-check", f"not Bash: {tool_name}", ctx=ctx
            )
            return

        tool_input = input_data.get("tool_input", {})
        command = tool_input.get("command", "")

        # Check if this is a gh pr create command
        if not is_pr_create_command(command):
            print_continue_and_log_skip("pr-issue-alignment-check", "not gh pr create", ctx=ctx)
            return

        # Extract issue numbers from body
        issue_numbers = extract_issue_numbers_from_body(command)
        if not issue_numbers:
            print_continue_and_log_skip(
                "pr-issue-alignment-check", "no issue numbers in body", ctx=ctx
            )
            return

        # Fetch and analyze issue content for each issue
        messages = []
        has_incomplete_open_criteria = False
        total_incomplete_open = 0
        closed_issues_with_incomplete = []

        for issue_num in issue_numbers:
            success, title, body, state = fetch_issue_content(issue_num)
            if not success:
                continue

            is_closed = state == "CLOSED"

            # Extract and check acceptance criteria
            criteria = extract_acceptance_criteria(body)

            if criteria:
                # Has acceptance criteria - format with completion status
                incomplete_count = sum(1 for is_completed, _ in criteria if not is_completed)
                if incomplete_count > 0:
                    if is_closed:
                        closed_issues_with_incomplete.append(issue_num)
                    else:
                        has_incomplete_open_criteria = True
                        total_incomplete_open += incomplete_count

                status_msg = format_acceptance_criteria_message(
                    issue_num, title, criteria, is_closed=is_closed
                )
                messages.append(status_msg)
            else:
                # No acceptance criteria - show issue content for reference
                display_body = body
                if len(display_body) > MAX_ISSUE_BODY_LENGTH:
                    display_body = display_body[:MAX_ISSUE_BODY_LENGTH] + "\n..."
                header = f"### Issue #{issue_num}: {title}"
                if is_closed:
                    header += " (CLOSED)"
                messages.append(f"{header}\n\n（受け入れ条件なし）\n\n{display_body}")

        if messages:
            if has_incomplete_open_criteria:
                # Strong warning for incomplete criteria on OPEN issues
                result["systemMessage"] = (
                    "🚨 **警告: 未完了の受け入れ条件があります！**\n\n"
                    f"❌ このPRでクローズされる全てのIssueの受け入れ条件のうち、"
                    f"合計 {total_incomplete_open} 件が未完了です。\n\n"
                    + "\n\n---\n\n".join(messages)
                    + "\n\n"
                    "⚠️ **このPRをマージすると、Issueが不完全な状態で"
                    "クローズされる可能性があります。**\n\n"
                    "確認してください:\n"
                    "1. 実装内容がIssueの全ての要求を満たしていますか？\n"
                    "2. 未完了の項目は意図的に対象外としていますか？\n"
                    "3. Issueの受け入れ条件を更新する必要がありますか？"
                )
                log_hook_execution(
                    "pr-issue-alignment-check",
                    "approve",
                    f"未完了の受け入れ条件あり: {total_incomplete_open}件 "
                    f"(#{', #'.join(issue_numbers)})",
                )
            elif closed_issues_with_incomplete:
                # Info message for closed issues with incomplete criteria
                result["systemMessage"] = (
                    "ℹ️ **PR作成前のIssue確認**\n\n" + "\n\n---\n\n".join(messages) + "\n\n"
                    f"💡 Issue #{', #'.join(closed_issues_with_incomplete)} は既に"
                    "クローズ済みのため、`Closes #N` は効果がありません。"
                )
                log_hook_execution(
                    "pr-issue-alignment-check",
                    "approve",
                    f"CLOSED Issueへの参照: #{', #'.join(closed_issues_with_incomplete)}",
                )
            else:
                # Info message when all criteria complete or no criteria
                result["systemMessage"] = (
                    "✅ **PR作成前のIssue確認**\n\n" + "\n\n---\n\n".join(messages) + "\n\n"
                    "💡 実装内容がIssueの要求と一致していることを確認してください。"
                )
                log_hook_execution(
                    "pr-issue-alignment-check",
                    "approve",
                    f"受け入れ条件確認: #{', #'.join(issue_numbers)}",
                )

    except Exception as e:
        # Don't block on errors
        log_hook_execution(
            "pr-issue-alignment-check",
            "error",
            f"フックエラー: {e}",
        )

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/pr_issue_alignment_check.ts
#!/usr/bin/env bun
/**
 * PR作成時に対象Issueの受け入れ条件を検証する。
 *
 * Why:
 *   - Issue #538 was closed with a PR that implemented something different
 *   - Issue #590 was closed with a PR that only added debug logs (not a fix)
 *   受け入れ条件を可視化し、不完全な状態でのクローズを防止する。
 *
 * What:
 *   - gh pr create コマンドを検出
 *   - Closes/Fixes キーワードからIssue番号を抽出
 *   - Issue内容と受け入れ条件を取得
 *   - 未完了の条件がある場合は警告
 *
 * Remarks:
 *   - 非ブロック型（警告のみ）
 *   - closes-validation.pyはキーワード検証、このフックは内容検証
 *   - Python版: pr_issue_alignment_check.py
 *
 * Changelog:
 *   - silenvx/dekita#543: フック追加
 *   - silenvx/dekita#592: 受け入れ条件チェック強化
 *   - silenvx/dekita#2917: TypeScript版初期実装
 */

import { execSync } from "node:child_process";
import { TIMEOUT_MEDIUM } from "../lib/constants";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { createHookContext, parseHookInput } from "../lib/session";
import { stripQuotedStrings } from "../lib/strings";

const HOOK_NAME = "pr-issue-alignment-check";
const MAX_ISSUE_BODY_LENGTH = 1000;

/**
 * Extract acceptance criteria (checkbox items) from Issue body.
 */
export function extractAcceptanceCriteria(body: string): Array<[boolean, string]> {
  const criteria: Array<[boolean, string]> = [];

  // Match checkbox items: - [ ] or - [x] or - [X]
  // Also handles * [ ] format
  const pattern = /^[\s]*[-*]\s*\[([ xX])\]\s*(.+)$/;

  for (const line of body.split("\n")) {
    const match = line.match(pattern);
    if (match) {
      const isCompleted = match[1].toLowerCase() === "x";
      const criteriaText = match[2].trim();
      criteria.push([isCompleted, criteriaText]);
    }
  }

  return criteria;
}

/**
 * Format status message for Issue acceptance criteria.
 */
export function formatAcceptanceCriteriaMessage(
  issueNum: string,
  title: string,
  criteria: Array<[boolean, string]>,
  isClosed = false,
): string {
  const incomplete = criteria.filter(([isCompleted]) => !isCompleted).map(([, text]) => text);
  const completedItems = criteria.filter(([isCompleted]) => isCompleted).map(([, text]) => text);

  let header = `### Issue #${issueNum}: ${title}`;
  if (isClosed) {
    header += " (CLOSED)";
  }

  const lines: string[] = [header, ""];

  if (isClosed && incomplete.length > 0) {
    lines.push("ℹ️ *このIssueは既にクローズ済みです。`Closes #N` は効果がありません。*");
    lines.push("");
  }

  if (incomplete.length > 0) {
    lines.push(`❌ **未完了の受け入れ条件: ${incomplete.length}件**`);
    for (const text of incomplete) {
      lines.push(`  - [ ] ${text}`);
    }
    lines.push("");
  }

  if (completedItems.length > 0) {
    lines.push(`✅ 完了済み: ${completedItems.length}件`);
    for (const text of completedItems) {
      lines.push(`  - [x] ${text}`);
    }
  }

  return lines.join("\n");
}

/**
 * Extract issue numbers from Closes/Fixes keywords in PR body.
 */
export function extractIssueNumbersFromBody(command: string): string[] {
  let body: string | null = null;

  // Try HEREDOC pattern first (most common in this project)
  // Matches: --body "$(cat <<'EOF' ... EOF )"
  const heredocMatch = command.match(/--body\s+"\$\(cat\s+<<['"]?EOF['"]?\s*(.*?)\s*EOF\s*\)"/s);
  if (heredocMatch) {
    body = heredocMatch[1];
  }

  // Try double-quoted body (may contain escaped quotes)
  if (body === null) {
    const dqMatch = command.match(/--body\s+"((?:[^"\\]|\\.)*)"/);
    if (dqMatch) {
      body = dqMatch[1];
    }
  }

  // Try single-quoted body (may contain any chars except single quote)
  if (body === null) {
    const sqMatch = command.match(/--body\s+'([^']*)'/);
    if (sqMatch) {
      body = sqMatch[1];
    }
  }

  if (body === null) {
    return [];
  }

  // Find Closes #XXX, Fixes #XXX, Resolves #XXX patterns
  // Case insensitive, handles multiple issues, allows optional colon
  const pattern = /(?:closes?|fix(?:es)?|resolves?):?\s+#(\d+)/gi;
  const matches: string[] = [];
  for (let match = pattern.exec(body); match !== null; match = pattern.exec(body)) {
    matches.push(match[1]);
  }

  return [...new Set(matches)]; // Remove duplicates
}

/**
 * Check if command is gh pr create.
 */
export function isPrCreateCommand(command: string): boolean {
  const cmd = stripQuotedStrings(command);
  return /gh\s+pr\s+create\b/.test(cmd);
}

/**
 * Fetch issue title, body and state using gh CLI.
 */
function fetchIssueContent(issueNumber: string): [boolean, string, string, string] {
  try {
    const result = execSync(`gh issue view ${issueNumber} --json title,body,state`, {
      encoding: "utf-8",
      timeout: TIMEOUT_MEDIUM * 1000,
      stdio: ["pipe", "pipe", "pipe"],
    });

    const data = JSON.parse(result);
    const title = data.title || "";
    const body = data.body || ""; // Handle null body
    const state = data.state || "OPEN";
    return [true, title, body, state];
  } catch {
    return [false, "", "", ""];
  }
}

interface ApproveResult {
  systemMessage?: string;
}

async function main(): Promise<void> {
  const result: ApproveResult = {};
  let sessionId: string | undefined;

  try {
    const data = await parseHookInput();
    const ctx = createHookContext(data);
    sessionId = ctx.sessionId;
    const toolName = (data.tool_name as string) || "";

    if (toolName !== "Bash") {
      await logHookExecution(HOOK_NAME, "approve", `not Bash: ${toolName}`, undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    const toolInput = (data.tool_input as Record<string, unknown>) || {};
    const command = (toolInput.command as string) || "";

    // Check if this is a gh pr create command
    if (!isPrCreateCommand(command)) {
      await logHookExecution(HOOK_NAME, "approve", "not gh pr create", undefined, { sessionId });
      console.log(JSON.stringify(result));
      return;
    }

    // Extract issue numbers from body
    const issueNumbers = extractIssueNumbersFromBody(command);
    if (issueNumbers.length === 0) {
      await logHookExecution(HOOK_NAME, "approve", "no issue numbers in body", undefined, {
        sessionId,
      });
      console.log(JSON.stringify(result));
      return;
    }

    // Fetch and analyze issue content for each issue
    const messages: string[] = [];
    let hasIncompleteOpenCriteria = false;
    let totalIncompleteOpen = 0;
    const closedIssuesWithIncomplete: string[] = [];

    for (const issueNum of issueNumbers) {
      const [success, title, body, state] = fetchIssueContent(issueNum);
      if (!success) {
        continue;
      }

      const isClosed = state === "CLOSED";

      // Extract and check acceptance criteria
      const criteria = extractAcceptanceCriteria(body);

      if (criteria.length > 0) {
        // Has acceptance criteria - format with completion status
        const incompleteCount = criteria.filter(([isCompleted]) => !isCompleted).length;
        if (incompleteCount > 0) {
          if (isClosed) {
            closedIssuesWithIncomplete.push(issueNum);
          } else {
            hasIncompleteOpenCriteria = true;
            totalIncompleteOpen += incompleteCount;
          }
        }

        const statusMsg = formatAcceptanceCriteriaMessage(issueNum, title, criteria, isClosed);
        messages.push(statusMsg);
      } else {
        // No acceptance criteria - show issue content for reference
        let displayBody = body;
        if (displayBody.length > MAX_ISSUE_BODY_LENGTH) {
          displayBody = `${displayBody.slice(0, MAX_ISSUE_BODY_LENGTH)}\n...`;
        }
        let header = `### Issue #${issueNum}: ${title}`;
        if (isClosed) {
          header += " (CLOSED)";
        }
        messages.push(`${header}\n\n（受け入れ条件なし）\n\n${displayBody}`);
      }
    }

    if (messages.length > 0) {
      if (hasIncompleteOpenCriteria) {
        // Strong warning for incomplete criteria on OPEN issues
        result.systemMessage = `🚨 **警告: 未完了の受け入れ条件があります！**\n\n❌ このPRでクローズされる全てのIssueの受け入れ条件のうち、合計 ${totalIncompleteOpen} 件が未完了です。\n\n${messages.join("\n\n---\n\n")}\n\n⚠️ **このPRをマージすると、Issueが不完全な状態でクローズされる可能性があります。**\n\n確認してください:\n1. 実装内容がIssueの全ての要求を満たしていますか？\n2. 未完了の項目は意図的に対象外としていますか？\n3. Issueの受け入れ条件を更新する必要がありますか？`;
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `未完了の受け入れ条件あり: ${totalIncompleteOpen}件 (#${issueNumbers.join(", #")})`,
          undefined,
          { sessionId },
        );
      } else if (closedIssuesWithIncomplete.length > 0) {
        // Info message for closed issues with incomplete criteria
        result.systemMessage = `ℹ️ **PR作成前のIssue確認**\n\n${messages.join("\n\n---\n\n")}\n\n💡 Issue #${closedIssuesWithIncomplete.join(", #")} は既にクローズ済みのため、\`Closes #N\` は効果がありません。`;
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `CLOSED Issueへの参照: #${closedIssuesWithIncomplete.join(", #")}`,
          undefined,
          { sessionId },
        );
      } else {
        // Info message when all criteria complete or no criteria
        result.systemMessage = `✅ **PR作成前のIssue確認**\n\n${messages.join("\n\n---\n\n")}\n\n💡 実装内容がIssueの要求と一致していることを確認してください。`;
        await logHookExecution(
          HOOK_NAME,
          "approve",
          `受け入れ条件確認: #${issueNumbers.join(", #")}`,
          undefined,
          { sessionId },
        );
      }
    }
  } catch (error) {
    // Don't block on errors
    await logHookExecution(HOOK_NAME, "error", `フックエラー: ${formatError(error)}`, undefined, {
      sessionId,
    });
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
